OpenRA.Mods.D2k/Traits/Sandworm.cs

```diff
@@ -0,0 +1,139 @@
+#region Copyright & License Information
+/*
+* Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+* This file is part of OpenRA, which is free software. It is made
+* available to you under the terms of the GNU General Public License
+* as published by the Free Software Foundation. For more information,
+* see COPYING.
+*/
+#endregion
+
+using System;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	class SandwormInfo : WandersInfo, Requires<MobileInfo>, Requires<RenderUnitInfo>, Requires<AttackBaseInfo>
+	{
+		[Desc("Time between rescanning for targets (in ticks).")]
+		public readonly int TargetRescanInterval = 32;
+
+		[Desc("The radius in which the worm \"searches\" for targets.")]
+		public readonly WRange MaxSearchRadius = WRange.FromCells(20);
+
+		[Desc("The range at which the worm launches an attack regardless of noise levels.")]
+		public readonly int IgnoreDistractionsAttackRange = 3;
+
+		[Desc("A number representing the chance this actor has of disappearing after it attacks (in %).")]
+		public readonly int ChanceToDisappear = 80;
+
+		[Desc("Name of the sequence that is used when the actor is idle or moving (not attacking).")]
+		public readonly string IdleSequence = "idle";
+
+		public override object Create(ActorInitializer init) { return new Sandworm(init.Self, this); }
+	}
+
+	class Sandworm : Wanders, ITick
+	{
+		public readonly SandwormInfo Info;
+
+		readonly Lazy<Mobile> mobile;
+		readonly Lazy<RenderUnit> renderUnit;
+		readonly Lazy<AttackBase> attackTrait;
+
+		public bool IsAttacking;
+		public bool IsMovingTowardTarget;
+
+		int targetCountdown;
+
+		public Sandworm(Actor self, SandwormInfo info)
+			: base(self, info)
+		{
+			Info = info;
+			mobile = Exts.Lazy(self.Trait<Mobile>);
+			renderUnit = Exts.Lazy(self.Trait<RenderUnit>);
+			attackTrait = Exts.Lazy(self.Trait<AttackBase>);
+		}
+
+		public override void OnBecomingIdle(Actor self)
+		{
+			if (renderUnit.Value.DefaultAnimation.CurrentSequence.Name != Info.IdleSequence)
+				renderUnit.Value.DefaultAnimation.ReplaceAnim("idle");
+
+			base.OnBecomingIdle(self);
+		}
+
+		public override void DoAction(Actor self, CPos targetCell)
+		{
+			IsMovingTowardTarget = false;
+
+			RescanForTargets(self);
+
+			if (IsMovingTowardTarget)
+				return;
+
+			Game.Debug("(Sandworm {0}) is wandering.", self.ActorID);
+			self.QueueActivity(mobile.Value.MoveWithinRange(Target.FromCell(self.World, targetCell, SubCell.Any), WRange.FromCells(1)));
+		}
+
+		public void Tick(Actor self)
+		{
+			if (--targetCountdown > 0 || IsAttacking || !self.IsInWorld)
+				return;
+
+			RescanForTargets(self);
+		}
+
+		void RescanForTargets(Actor self)
+		{
+			targetCountdown = Info.TargetRescanInterval;
+
+			var actorsInRange = self.World.FindActorsInCircle(self.CenterPosition, Info.MaxSearchRadius);
```

Comment:
Because you select your target from within this set, the sandworms don't actually need the `RevealsShroud` trait anymore, so you can get a little perf back.